File: DSA/Graph/Code/CycleDetection.cpp
#include <iostream>
#include <vector>
using namespace std;

class CycleDetection {
private:
    int vertices;
    vector<vector<int>> adjacencyMatrix;
    
    bool hasCycleDFS(int u, int parent, vector<bool>& visited) {
        visited[u] = true;
        
        for (int v = 0; v < vertices; v++) {
            if (adjacencyMatrix[u][v]) {
                if (!visited[v]) {
                    if (hasCycleDFS(v, u, visited)) {
                        return true;
                    }
                } else if (v != parent) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
public:
    CycleDetection(int v) : vertices(v) {
        adjacencyMatrix.assign(v, vector<int>(v, 0));
    }
    
    void addEdge(int u, int v) {
        adjacencyMatrix[u][v] = 1;
        adjacencyMatrix[v][u] = 1;
    }
    
    bool detectCycle() {
        vector<bool> visited(vertices, false);
        
        for (int i = 0; i < vertices; i++) {
            if (!visited[i]) {
                if (hasCycleDFS(i, -1, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
};

int main() {
    cout << "=== Cycle Detection (C++) ===\n\n";
    
    cout << "Graph 1 (With Cycle):\n";
    CycleDetection graph1(4);
    graph1.addEdge(0, 1);
    graph1.addEdge(1, 2);
    graph1.addEdge(2, 3);
    graph1.addEdge(3, 0);
    
    cout << "Edges: 0-1, 1-2, 2-3, 3-0 (forms a cycle)\n";
    cout << "Cycle Detected: " << (graph1.detectCycle() ? "Yes" : "No") << "\n\n";
    
    cout << "Graph 2 (Acyclic):\n";
    CycleDetection graph2(4);
    graph2.addEdge(0, 1);
    graph2.addEdge(1, 2);
    graph2.addEdge(2, 3);
    
    cout << "Edges: 0-1, 1-2, 2-3 (no cycle)\n";
    cout << "Cycle Detected: " << (graph2.detectCycle() ? "Yes" : "No") << "\n\n";
    
    cout << "=== Algorithm ===\n";
    cout << "1. Use DFS to traverse graph\n";
    cout << "2. For each vertex, track parent\n";
    cout << "3. If visited neighbor is not parent = cycle found\n";
    cout << "4. Continue for all components\n\n";
    
    cout << "=== Complexity ===\n";
    cout << "Time Complexity:  O(V + E)\n";
    cout << "Space Complexity: O(V)\n";
    
    return 0;
}

File: DSA/Graph/Code/CycleDetection.py
# Cycle Detection in Python

class CycleDetection:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency_matrix = [[0] * vertices for _ in range(vertices)]
    
    def add_edge(self, u, v):
        self.adjacency_matrix[u][v] = 1
        self.adjacency_matrix[v][u] = 1
    
    def has_cycle_dfs(self, u, parent, visited):
        visited[u] = True
        
        for v in range(self.vertices):
            if self.adjacency_matrix[u][v]:
                if not visited[v]:
                    if self.has_cycle_dfs(v, u, visited):
                        return True
                elif v != parent:
                    return True
        
        return False
    
    def detect_cycle(self):
        visited = [False] * self.vertices
        
        for i in range(self.vertices):
            if not visited[i]:
                if self.has_cycle_dfs(i, -1, visited):
                    return True
        
        return False

# Test
print("=== Cycle Detection (Python) ===\n")

print("Graph 1 (With Cycle):")
graph1 = CycleDetection(4)
graph1.add_edge(0, 1)
graph1.add_edge(1, 2)
graph1.add_edge(2, 3)
graph1.add_edge(3, 0)

print("Edges: 0-1, 1-2, 2-3, 3-0 (forms a cycle)")
print(f"Cycle Detected: {'Yes' if graph1.detect_cycle() else 'No'}\n")

print("Graph 2 (Acyclic):")
graph2 = CycleDetection(4)
graph2.add_edge(0, 1)
graph2.add_edge(1, 2)
graph2.add_edge(2, 3)

print("Edges: 0-1, 1-2, 2-3 (no cycle)")
print(f"Cycle Detected: {'Yes' if graph2.detect_cycle() else 'No'}\n")

print("=== Algorithm ===")
print("1. Use DFS to traverse graph")
print("2. For each vertex, track parent")
print("3. If visited neighbor is not parent = cycle found")
print("4. Continue for all components\n")

print("=== Complexity ===")
print("Time Complexity:  O(V + E)")
print("Space Complexity: O(V)")
